### PR TITLE
Enable `disallow_untyped_defs` on mypy

### PR DESCRIPTION
**Pull request by johnnv1**
Enable `disallow_untyped_defs` on mypy and add the missing typing

### CODE REVIEW

**Comment by johnnv1** [comment]
enable https://beta.ruff.rs/docs/rules/#flake8-annotations-ann at ruff config

**Review by edgarriba — APPROVED** [review]

File: kornia/feature/sold2/backbones.py [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
+from kornia.core import Module, Tensor
+
+
+class HourglassConfig(NamedTuple):
+    depth: int
+    num_stacks: int
+    num_blocks: int
+    num_classes: int
+    input_channels: int
+    head: Type[Module]
```
Comment (johnnv1) [inline kornia/feature/sold2/backbones.py:18]: Some changes like this one, will need to be done -- related to #2093

File: kornia/core/tensor_wrapper.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+# type: ignore
```
Comment (johnnv1) [inline kornia/core/tensor_wrapper.py:1]: It needs a specific issue to type this file (I can't figure out how to do this).

Also, needs to remove the `# type: ignore[attr-defined]` from the files which imports the tensor wrapper

Comment (johnnv1) [reply kornia/core/tensor_wrapper.py:1]: I added this import type ignore on the files:
- kornia/geometry/plane.py
- kornia/geometry/vector.py

Comment (edgarriba) [reply kornia/core/tensor_wrapper.py:1]: @cclauss

File: kornia/augmentation/_2d/mix/base.py [hunk 3]
```diff
@@ -187,9 +187,9 @@ def forward(  # type: ignore[override]
         return outputs
 
     @torch.jit.ignore
-    def inverse(self, **kwargs):
+    def inverse(self, **kwargs: Any) -> Optional[Tensor]:
```
Comment (edgarriba) [inline kornia/augmentation/_2d/mix/base.py:190]:
```suggestion
    def inverse(self, **kwargs: Any) -> Tensor | None:
```

File: kornia/contrib/vit.py [hunk 4]
```diff
@@ -212,7 +212,7 @@ def __init__(
         self.encoder = TransformerEncoder(hidden_dim, depth, num_heads, dropout_rate, dropout_attn)
 
     @property
-    def encoder_results(self):
+    def encoder_results(self) -> List[Tensor]:
```
Comment (edgarriba) [inline kornia/contrib/vit.py:215]:
```suggestion
    def encoder_results(self) -> list[Tensor]:
```

**Review by edgarriba — APPROVED** [review]
LTGM. There are some old style typings that can be isolated into a new issue

File: kornia/contrib/vit.py (hunk 4 above)
Comment (cclauss) [reply kornia/contrib/vit.py:215]: I believe that those conversions can be automated via https://beta.ruff.rs/docs/rules/#pyupgrade-up

File: kornia/core/tensor_wrapper.py (hunk 2 above)
Comment (cclauss) [reply kornia/core/tensor_wrapper.py:1]: I would be OK with @johnnv1 proposed approach to land these changes with line 1 as is as long as a new issue is opened to get rid of it.

The way to do this (in a separate PR) is to remove this comment and then run `mypy` (or another typing tool) and see what type errors are raised and then specify `# type: ignore[your-issue-here]` on individual lines.

**Comment by johnnv1** [comment]
> LTGM. There are some old style typings that can be isolated into a new issue

yes, to avoid removing jit support in this patch, I just added the new typing where jit was no longer supported. Later, when we get rid of jit, pyupgrade does it for us (via ruff)

**Pull request merged**